llama-cpp.nix

```diff
@@ -0,0 +1,118 @@
+{ lib
+, stdenv
+, cmake
+, ninja
+, pkg-config
+, symlinkJoin
+, python3
+, openmpi
+, openblas
+, cudaPackages
+, rocmPackages
+, clblast
+, Accelerate ? null
+, MetalKit ? null
+, CoreVideo ? null
+, CoreGraphics ? null
+, useOpenCL ? false
+, useCuda ? false
+, useRocm ? false
+}:
+
+let
+  inherit (lib) optional optionals versionOlder;
+  isDefault = !useOpenCL && !useCuda && !useRocm;
+
+  # Give a little description difference between the flavors.
+  descriptionSuffix = if useOpenCL then
+    " (OpenCL accelerated)"
+  else if useCuda then
+    " (CUDA accelerated)"
+  else if useRocm then
+    " (ROCm accelerated)"
+  else if (MetalKit != null) then
+    " (MetalKit accelerated)"
+  else "";
+
+  llama-python = python3.withPackages (ps: [ ps.numpy ps.sentencepiece ]);
+
+  # See ./overlay.nix for where these dependencies are passed in.
+  defaultBuildInputs = builtins.filter (p: p != null) [
+    Accelerate
+    MetalKit
+    CoreVideo
+    CoreGraphics
+  ];
+
+  cudaBuildInput = symlinkJoin (with cudaPackages; {
+    # HACK(Green-Sky): Nix currently has issues with CMake findcudatoolkit
+    # see https://github.com/NixOS/nixpkgs/issues/224291
+    # copied from jaxlib
+    name = "${cudatoolkit.name}-merged";
+    paths = [
+      cudatoolkit.lib
+      cudatoolkit.out
+    ] ++ optionals (versionOlder cudatoolkit.version "11") [
+      # For some reason, some of the required libs are in the targets/x86_64-linux
+      # directory; not sure why, but this works around it.
+      "${cudatoolkit}/targets/${stdenv.system}"
+    ];
+  });
+
+  rocmBuildInputs = with rocmPackages; [ clr hipblas rocblas ];
+in
+
+stdenv.mkDerivation {
+  name = "llama.cpp";
+  src = ./.;
+  meta = {
+    description = "Inference of LLaMA model in pure C/C++${descriptionSuffix}";
+    mainProgram = "llama";
+  };
+
+  postPatch = ''
+    substituteInPlace ./ggml-metal.m \
+      --replace '[bundle pathForResource:@"ggml-metal" ofType:@"metal"];' "@\"$out/bin/ggml-metal.metal\";"
```

Comment:
I'm not familiar with this, is `[bundle pathForResource:@"default" ofType:@"metallib"];` some standard syntax? Is it only with the nix package that we have to override this, or all users have to do that? In the latter case, could we address this at the source level? E.g. make it so this could be specified in the cmake options (`cmakeFlags`)?

Reply:
I'm also not familiar with what this is doing. I ported it over wholesale from @Green-Sky's original `flake.nix`.